# Numbers and more in Python!

In this lecture, we will learn about numbers in Python and how to use them.

We'll learn about the following topics:

    1.) Types of Numbers in Python
    2.) Basic Arithmetic
    3.) Differences between classic division and floor division
    4.) Object Assignment in Python

## Types of numbers

Python has various "types" of numbers (numeric literals). We'll mainly focus on integers and floating point numbers.

Integers are just whole numbers, positive or negative. For example: 2 and -2 are examples of integers.

Floating point numbers in Python are notable because they have a decimal point in them, or use an exponential (e) to define the number. For example 2.0 and -2.1 are examples of floating point numbers. 4E2 (4 times 10 to the power of 2) is also an example of a floating point number in Python.

Throughout this course we will be mainly working with integers or simple float number types.

Here is a table of the two main types we will spend most of our time working with some examples:

<table>
<tr>
    <th>Examples</th> 
    <th>Number "Type"</th>
</tr>

<tr>
    <td>1,2,-5,1000</td>
    <td>Integers</td> 
</tr>

<tr>
    <td>1.2,-0.5,2e2,3E2</td> 
    <td>Floating-point numbers</td> 
</tr>
 </table>

## BASIC ARITHMETIC

In [1]:
#addition
2+5

7

In [2]:
#subtraction
10-12

-2

In [3]:
#multiplication
2*4

8

In [6]:
#division
7/4

1.75

In [7]:
#floor division
7//4

1

**Whoa! What just happened? Last time I checked, 7 divided by 4 equals 1.75 not 1!**

The reason we get this result is because we are using "*floor*" division. The // operator (two forward slashes) truncates the decimal without rounding, and returns an integer result.

In [8]:
#modulo
7%5

2

In [9]:
#powers
3**2

9

In [10]:
#can you find the cube root of 125 using the operators mentioned above
125**(1/3)

5.0

## Variable Assignments

Now that we've seen how to use numbers in Python as a calculator let's see how we can assign names and create variables.

We use a single equals sign to assign labels to variables. Let's see a few examples of how we can do this.

The names you use when creating these labels need to follow a few rules:

    1. Names can not start with a number.
    2.Variable name can start with a letter or _
    3. Can't use any special characters except _
    4. It's considered best practice (PEP8) that names are lowercase.
    5. Avoid using words that have special meaning in Python like "list" and "str"


## Assigning Variables
Variable assignment follows `name = object`, where a single equals sign `=` is an *assignment operator*

In [12]:
name="Alan"
name

'Alan'

In [14]:
age=10
age

10

In [17]:
condition= True
condition

True

In [20]:
tax=123.45
tax

123.45

## Determining variable type with `type()`
You can check what type of object is assigned to a variable using Python's built-in `type()` function. Common data types include:
* **int** (for integer)
* **float**
* **str** (for string)
* **list**
* **tuple**
* **dict** (for dictionary)
* **set**
* **bool** (for Boolean True/False)

In [21]:
d={'Name':'Alen','Age':10}
type(d)

dict

In [22]:
ans=(2/3)**5

In [24]:
type(ans)

float

# Operators in Python
  
  The different types of operators in Python are:
  * Arithmetic operators
  * Comparison operators
  * Logical operators
  * Bitwise operators
  * Assignment operators
  * Identity operators
  * Membership operators

# Comparison Operators 

In this lecture we will be learning about Comparison Operators in Python. These operators will allow us to compare variables and output a Boolean value (True or False). 

If you have any sort of background in Math, these operators should be very straight forward.

First we'll present a table of the comparison operators and then work through some examples:

<h2> Table of Comparison Operators </h2>

<table class="table table-bordered">
<tr>
<th style="width:10%">Operator</th><th style="width:45%">Description</th><th>Example</th>
</tr>
<tr>
<td>==</td>
<td>If the values of two operands are equal, then the condition becomes true.</td>
<td> (a == b) is not true.</td>
</tr>
<tr>
<td>!=</td>
<td>If values of two operands are not equal, then condition becomes true.</td>
<td>(a != b) is true</td>
</tr>
<tr>
<td>&gt;</td>
<td>If the value of left operand is greater than the value of right operand, then condition becomes true.</td>
<td> (a &gt; b) is not true.</td>
</tr>
<tr>
<td>&lt;</td>
<td>If the value of left operand is less than the value of right operand, then condition becomes true.</td>
<td> (a &lt; b) is true.</td>
</tr>
<tr>
<td>&gt;=</td>
<td>If the value of left operand is greater than or equal to the value of right operand, then condition becomes true.</td>
<td> (a &gt;= b) is not true. </td>
</tr>
<tr>
<td>&lt;=</td>
<td>If the value of left operand is less than or equal to the value of right operand, then condition becomes true.</td>
<td> (a &lt;= b) is true. </td>
</tr>
</table>

In [25]:
a,b=3,4
print("a==b is",a==b)

a==b is False


In [26]:
print("a!=b is ",a!=b)

a!=b is  True


In [27]:
print("a>b is ",a>b)

a>b is  False


In [28]:
print("a>=b is ",a>=b)

a>=b is  False


In [29]:
print("a<b is ",a<b)

a<b is  True


In [30]:
print("a<=b is ",a<=b)

a<=b is  True


# Chained Comparison Operators

An interesting feature of Python is the ability to *chain* multiple comparisons to perform a more complex test. You can use these chained comparisons as shorthand for larger Boolean Expressions.

In this lecture we will learn how to chain comparison operators and we will also introduce two other important statements in Python: **and** and **or**.

Let's look at a few examples of using chains:

In [31]:
1< 3 >2

True

# Logical operators

<h2> Table of Logical Operators </h2>

<table class="table table-bordered">
<tr>
<th style="width:10%">Operator</th><th style="width:45%">Description</th><th>Example</th>
</tr>
<tr>
    <td>and</td>
    <td>True if both the operands are true</td>
    <td>a and b</td>
</tr>
    <tr>
    <td>or</td>
    <td>True if either of the operands are true</td>
    <td>a or b</td>
</tr>
<tr>
    <td>not</td>
    <td>Complements the operand</td>
    <td>not a</td>
</tr>
</table>

In [32]:
True and False

False

In [33]:
True or False

True

In [34]:
not False

True

In [35]:
2 and 3 and 4

4

In [36]:
2 or 3 or 4

2

# Bitwise Operators

Bitwise operators act on operands as if they were string of binary digits. It operates bit by bit, hence the name.
For example, 2 is 10 in binary and 7 is 111.

<h2>Table of Bitwise operators</h2>
<table class="table table-bordered">
    <tr><th style="width:10%">Operator</th><th style="width=45%">Description</th><th>Example</th></tr>
    <tr>
        <td>&</td>
        <td>Bitwise AND</td>
        <td>a & b</td>
    </tr>
    <tr>
        <td>|</td>
        <td>Bitwise OR</td>
        <td>a | b</td>
    </tr>
    <tr>
        <td>~</td>
        <td>Bitwise NOT</td>
        <td>~a</td>
    </tr>
    <tr>
        <td>^</td>
        <td>Bitwise XOR</td>
        <td>a ^ b</td>
    </tr>
    <tr>
        <td>>></td>
        <td>Bitwise Right shift</td>
        <td>a >> b</td>
    </tr>
    <tr>
        <td> << </td>
        <td>Bitwise Left Shift</td>
        <td>a << b</td>
    </tr>
</table>

In [37]:
a,b=2,7

In [38]:
a & b

2

In [39]:
a | b

7

In [40]:
~a

-3

In [41]:
a ^ b

5

In [42]:
a >> b

0

In [43]:
a << b

256

# Assignment Operators

Assignment operators are used in Python to assign values to variables.`a = 5` is a simple assignment operator that assigns the value 5 on the right to the variable a on the left.There are various compound operators in      Python like    `a += 5` that adds to the variable and later assigns the same. It is equivalent to `a = a + 5`.

<h2>Table of Assignment operators</h2>
<table class="table table-bordered">
    <tr><th style="width:10%">Operator</th><th style="width:20%">Description</th></tr>
    <tr>
        <td>=</td>
        <td> a=10</td>
    </tr>
    <tr>
        <td>+=</td>
        <td>a+=10</td>
    </tr>
    <tr>
        <td>-=</td>
        <td>a-=10</td>
    </tr>
    <tr>
        <td>*=</td>
        <td>a*=10</td>
    </tr>
    <tr>
        <td>/=</td>
        <td>a/=10</td>
    </tr>
    <tr>
        <td>%=</td>
        <td>a%=10</td>
    </tr>
    <tr>
        <td>//=</td>
        <td>a//=10</td>
    </tr>
    <tr>
        <td>**=</td>
        <td>a**=10</td>
    </tr>
    <tr>
        <td>&=</td>
        <td>a&=10</td>
    </tr>
    <tr>
        <td>|=</td>
        <td>a|=10</td>
    </tr>
    <tr>
        <td>^=</td>
        <td>a^=10</td>
    </tr>
    <tr>
        <td>>>=</td>
        <td>a>>=10</td>
    </tr>
    <tr>
        <td><<=</td>
        <td>a<<=10</td>
    </tr>
</table>